### PR TITLE
Protocol async errors refactor

### PR DESCRIPTION
**Pull request by pradovic**
Refactor all async handlers that use `p2p.Protocol.Run()` to not log errors in handlers, and not to use `peer.Drop` directly.

todo: 
- [x] Add special error to allow handlers to return error to be logged but not break the execution. Identified places are commented with `todo: handle nil error`
- [ ] There are a few more "double async" parts in a few already async handlers that still use `peer.Drop()`. Investigate and remove if possible (will be done in separate PR)

I have link this PR to https://github.com/ethersphere/swarm/pull/2058 at the moment, but I will merge them separately on master later. I did this as they are chronologically related.

### CODE REVIEW

File: network/stream/peer.go [hunk 1]
```diff
@@ -134,32 +134,29 @@ type want struct {
...
+// in case the offer is not found - error is returned
+func (p *Peer) getOffer(ruid uint) (o offer, err error) {
 	p.mtx.RLock()
 	o, ok := p.openOffers[ruid]
 	p.mtx.RUnlock()
 	if !ok {
-		p.logger.Error("ruid not found, dropping peer", "ruid", ruid)
-		p.Drop("ruid not found")
-		return o, true
+		return o, fmt.Errorf("ruid not found, dropping peer: %d", ruid)
```
Comment (janos) [inline network/stream/peer.go]: Maybe to remove `dropping` from the error message, as that consequence is handled in different place.

File: network/stream/peer.go [hunk 2]
```diff
@@ -134,32 +134,29 @@ type want struct {
...
+// in case the want is not found the error is returned
+func (p *Peer) getWant(ruid uint) (w *want, err error) {
 	p.mtx.RLock()
 	w, ok := p.openWants[ruid]
 	p.mtx.RUnlock()
 	if !ok {
-		p.logger.Error("ruid not found, dropping peer", "ruid", ruid)
-		p.Drop("ruid not found")
-		return nil, true
+		return nil, fmt.Errorf("ruid not found, dropping peer: %d", ruid)
```
Comment (janos) [inline network/stream/peer.go]: Maybe to remove `dropping` from the error message, as that consequence is handled in different place.

File: network/stream/stream.go [hunk 3]
```diff
@@ -392,7 +358,7 @@ func (r *Registry) serverHandleGetRange(ctx context.Context, p *Peer, msg *GetRa
 
 	key, err := provider.ParseKey(msg.Stream.Key)
 	if err != nil {
-		return fmt.Errorf("erroring parsing stream key: %w", err)
+		return fmt.Errorf("erroring parsing stream key, stream: %s, err: %w", msg.Stream, err)
```
Comment (janos) [inline network/stream/stream.go]: Word `erroring` is not needed, as it is clear that the error is returned. I am suggesting something like `parse stream key...`. This goes to all similar errors, I know that you did not write them, but it is a nice opportunity to fix them.

File: network/stream/stream.go [hunk 4]
```diff
@@ -456,14 +426,23 @@ func (r *Registry) serverHandleGetRange(ctx context.Context, p *Peer, msg *GetRa
 		p.mtx.Lock()
 		delete(p.openOffers, msg.Ruid)
 		p.mtx.Unlock()
-		return fmt.Errorf("error sending offered hashes: %w", err)
+		return fmt.Errorf("error sending offered hashes, ruid: %d, err: %w", msg.Ruid, err)
```
Comment (janos) [inline network/stream/stream.go]: `error sending offered...` -> `send offered...`

File: network/stream/stream.go [hunk 5]
```diff
@@ -489,8 +468,7 @@ func (r *Registry) clientHandleOfferedHashes(ctx context.Context, p *Peer, msg *
 	// the LastIndex on the incoming message. we should seal the interval and request the subsequent
 	if lenHashes == 0 {
 		if err := p.sealWant(w); err != nil {
-			p.logger.Error("error persisting interval", "from", w.from, "to", w.to, "err", err)
-			return fmt.Errorf("error persisting interval: %w", err)
+			return fmt.Errorf("error persisting interval: from %d, to: %d, err: %w", w.from, w.to, err)
```
Comment (janos) [inline network/stream/stream.go]: This error message is not related to what the function `sealWant` suggests to do. I know that this is not the message you wrote, but it should be fixed.

File: network/stream/stream.go [hunk 6]
```diff
@@ -504,12 +482,14 @@ func (r *Registry) clientHandleOfferedHashes(ctx context.Context, p *Peer, msg *
 	// the upstream peer in order to mitigate a leak on `offer`s
 	if !provider.WantStream(p, w.stream) {
 		wantedHashesMsg.BitVector = []byte{}
-		return p.Send(ctx, wantedHashesMsg)
+		if err := p.Send(ctx, wantedHashesMsg); err != nil {
+			return fmt.Errorf("error sending empty wanted hashes:  %w", err)
```
Comment (janos) [inline network/stream/stream.go]: `error sending empty wanted hashes...` -> `send empty wanted hashes...`

File: swap/swap.go [hunk 7]
```diff
@@ -405,17 +403,18 @@ func (s *Swap) handleEmitChequeMsg(ctx context.Context, p *Peer, msg *EmitCheque
...
 	if err != nil {
-		log.Error("error getting contract", "err", err)
-		return err
+		return fmt.Errorf("error getting contract: %w", err)
 	}
 
 	gasPrice, err := s.backend.SuggestGasPrice(context.TODO())
 	if err != nil {
-		return err
+		// todo: handle nil error
```
Comment (janos) [inline swap/swap.go]: It is better to log the error here if it is not propagated.

File: swap/swap.go [hunk 8]
```diff
@@ -405,17 +403,18 @@ func (s *Swap) handleEmitChequeMsg(ctx context.Context, p *Peer, msg *EmitCheque
...
+		return nil
 	}
 	transactionCosts := gasPrice.Uint64() * 50000 // cashing a cheque is approximately 50000 gas
 	paidOut, err := otherSwap.PaidOut(nil, cheque.Beneficiary)
 	if err != nil {
+		// todo: handle nil error
```
Comment (janos) [inline swap/swap.go]: It is better to log the error here if it is not propagated.

File: swap/swap.go [hunk 9]
```diff
@@ -426,35 +425,37 @@ func (s *Swap) handleEmitChequeMsg(ctx context.Context, p *Peer, msg *EmitCheque
 		go defaultCashCheque(s, otherSwap, opts, cheque)
 	}
 
-	return err
+	// todo: handle nil error
+	return nil
```
Comment (janos) [inline swap/swap.go:446]: It is better to log the error here if it is not propagated.

**Review by janos — COMMENTED** [review]
It looks good to me in important parts. It seems like a nice opportunity to fix error messages that are there from start like that. I did comment on some of them, but then I saw that they are all badly written, so could you adjust them, so that they do not contain work `error` or `erroring` as it is clear that they are errors.

File: swap/swap.go (hunk 7 above)
Comment (pradovic) [reply swap/swap.go]: I agree. The idea was to introduce a special error that is used when the loop is to be finished, and to just return error in the cases like this so it can be logged by `protocol.Run()`. Do you think it makes sense or should we just log in the handlers and return nil?

Comment (pradovic) [reply swap/swap.go]: Basically the usage difference would be:
1. In order to log an error, but not break the loop, return err
2. In ordeer to log and error and break the loop, use special error

Or we could just log in the handler, then the usage would be:
1. In order to log an error and break the loop -  return err
2. In order to log and error but not break the loop - log err in handler and return nil

Comment (janos) [reply swap/swap.go]: It makes sense. If you would like to leave new error type for another PR, then it is ok to just log the error here (in this PR), as this is its last handling, for now.

Comment (pradovic) [reply swap/swap.go]: Sounds good, thanks!

Comment (janos) [reply swap/swap.go]: It would be good to hear @zelig's opinion also. I think that DropError type which encapsulates original error would make sense, as we could have also BlacklistError in a similar manner, but would contain additional information about blacklisting period or other information.

Comment (pradovic) [reply swap/swap.go]: Makes sense to me as well. I will just log an error here then at this point, and will open a separate PR for `DropError` then, and we can discuss there.

Comment (pradovic) [reply swap/swap.go]: I ended up doing it in the same PR, as it is part of the similar functionality, and I wanted to avoid you guys needing to go through very similar code and PRs twice, especially chronologically related. I hope that is good :)

**Comment by pradovic** [comment]
@janos , @zelig , I have updated PR with the idea how to handle different kind of errors. I have removed the old errors in `p2p.Protocol` as I did not really see them used anywhere logically, just for error messages. 
Basically the idea is that special errors inside the handlers and p2p.Protocol define how they can be handled logically (should they break the loop, etc...), while handling them and logging additional context in the specific protocols present the information to the user. This whey we can add more specific errors, and log differently several use-cases, for ex. graceful shutdown, fatal error, blacklisting, etc... 
 Let me know do you think that approach makes sense, and if you like, we can brainstorm a bit about the log messages format on a different levels of protocol.

File: bzzeth/bzzeth.go [hunk 10]
```diff
@@ -135,8 +143,7 @@ func (b *BzzEth) handleNewBlockHeaders(ctx context.Context, p *Peer, msg *NewBlo
 	}
 	yes, err := b.netStore.Store.HasMulti(ctx, addresses...)
 	if err != nil {
-		log.Error("Error checking hashesh in store", "Reason", err)
-		return nil
+		return fmt.Errorf("error checking hashesh in store. Reason: %w", err)
```
Comment (janos) [inline bzzeth/bzzeth.go]: I do not think that `error` and `Reason` is needed in message, Maybe something like `check hashes: %w`?

File: bzzeth/bzzeth.go [hunk 11]
```diff
@@ -79,6 +79,7 @@ func (b *BzzEth) Run(p *p2p.Peer, rw p2p.MsgReadWriter) error {
 	// perform handshake and register if peer serves headers
 	handshake, err := bp.Handshake(context.TODO(), Handshake{ServeHeaders: true}, nil)
 	if err != nil {
+		log.Error("Shutting down bzzeth protocol.", "peer", bp, "reason", err)
```
Comment (janos) [inline bzzeth/bzzeth.go]: Error should not be logged and passed, just one of those.

File: bzzeth/bzzeth.go [hunk 12]
```diff
@@ -90,10 +91,17 @@ func (b *BzzEth) Run(p *p2p.Peer, rw p2p.MsgReadWriter) error {
...
 	// If another swarm node tries to connect then the protocol goes into idle
 	if isSwarmNodeFunc(bp) {
-		return peer.Run(b.handleMsgFromSwarmNode(bp))
+		err = peer.Run(b.handleMsgFromSwarmNode(bp))
 	}
 
-	return peer.Run(b.handleMsg(bp))
+	err = peer.Run(b.handleMsg(bp))
+	if err != nil {
+		log.Error("Shutting down bzzeth protocol.", "peer", bp, "reason", err)
```
Comment (janos) [inline bzzeth/bzzeth.go]: Error should not be logged and passed, just one of those.

File: bzzeth/bzzeth.go [hunk 13]
```diff
@@ -90,10 +91,17 @@ func (b *BzzEth) Run(p *p2p.Peer, rw p2p.MsgReadWriter) error {
 	// This protocol is all about interaction between an Eth node and a Swarm Node.
 	// If another swarm node tries to connect then the protocol goes into idle
 	if isSwarmNodeFunc(bp) {
-		return peer.Run(b.handleMsgFromSwarmNode(bp))
+		err = peer.Run(b.handleMsgFromSwarmNode(bp))
```
Comment (janos) [inline bzzeth/bzzeth.go]: This error is shadowed later.

File: bzzeth/bzzeth.go [hunk 14]
```diff
@@ -118,7 +126,7 @@ func (b *BzzEth) handleMsg(p *Peer) func(context.Context, interface{}) error {
 func (b *BzzEth) handleMsgFromSwarmNode(p *Peer) func(context.Context, interface{}) error {
 	return func(ctx context.Context, msg interface{}) error {
 		p.logger.Warn("bzzeth.handleMsgFromSwarmNode")
-		return errRcvdMsgFromSwarmNode
+		return protocols.BreakError(errRcvdMsgFromSwarmNode)
```
Comment (janos) [inline bzzeth/bzzeth.go]: Warning should be removed, even more as it does not add any context to the error.

File: bzzeth/bzzeth.go [hunk 15]
```diff
@@ -160,8 +167,7 @@ func (b *BzzEth) handleNewBlockHeaders(ctx context.Context, p *Peer, msg *NewBlo
 	deliveries := make(chan []byte)
 	req, err := p.getBlockHeaders(ctx, hashes, deliveries)
 	if err != nil {
-		p.logger.Error("Error sending GetBlockHeader message", "Reason", err)
-		return nil
+		return fmt.Errorf("error sending GetBlockHeader message. Reason: %w", err)
```
Comment (janos) [inline bzzeth/bzzeth.go]: Similar like in an error a few lines above.

File: network/retrieval/retrieve.go [hunk 16]
```diff
@@ -147,7 +147,13 @@ func (r *Retrieval) Run(bp *network.BzzPeer) error {
 	r.addPeer(sp)
 	defer r.removePeer(sp)
 
-	return sp.Peer.Run(r.handleMsg(sp))
+	if err := sp.Peer.Run(r.handleMsg(sp)); err != nil {
+		log.Error("Shutting down retrieve protocol.", "peer", sp.ShortString(), "reason", err)
```
Comment (janos) [inline network/retrieval/retrieve.go]: Error should not be logged and returned.

File: network/retrieval/retrieve.go [hunk 17]
```diff
@@ -337,10 +341,7 @@ func (r *Retrieval) handleRetrieveRequest(ctx context.Context, p *Peer, msg *Ret
 
 	err = p.Send(ctx, deliveryMsg)
 	if err != nil {
-		p.logger.Error("retrieval.handleRetrieveRequest - peer delivery failed", "ref", msg.Addr, "err", err)
-		osp.LogFields(olog.Bool("delivered", false))
-		// continue in event loop
-		return nil
+		return fmt.Errorf("retrieval.handleRetrieveRequest - peer delivery failed. ref: %s, err: %w", msg.Addr, err)
```
Comment (janos) [inline network/retrieval/retrieve.go]: No need for `err:` in message, just `:`.

File: network/retrieval/retrieve.go [hunk 18]
```diff
@@ -385,10 +385,11 @@ func (r *Retrieval) handleChunkDelivery(ctx context.Context, p *Peer, msg *Chunk
 
 	_, err = r.netStore.Put(ctx, mode, storage.NewChunk(msg.Addr, msg.SData))
 	if err != nil {
-		p.logger.Error("netstore error putting chunk to localstore", "err", err)
 		if err == storage.ErrChunkInvalid {
-			return fmt.Errorf("netstore error putting chunk to localstore: %s", err)
+			return protocols.BreakError(fmt.Errorf("netstore error putting chunk to localstore. err: %w", err))
```
Comment (janos) [inline network/retrieval/retrieve.go]: No need for `. err:` in message, just `:`.

File: network/retrieval/retrieve.go [hunk 19]
```diff
@@ -385,10 +385,11 @@ func (r *Retrieval) handleChunkDelivery(ctx context.Context, p *Peer, msg *Chunk
...
 		}
+
+		return fmt.Errorf("netstore error putting chunk to localstore. err: %w", err)
```
Comment (janos) [inline network/retrieval/retrieve.go]: No need for `. err:` in message, just `:`.

File: p2p/protocols/error.go [hunk 20]
```diff
@@ -0,0 +1,22 @@
+package protocols
```
Comment (janos) [inline p2p/protocols/error.go:17]: Add file header.

File: p2p/protocols/error.go [hunk 21]
```diff
@@ -0,0 +1,22 @@
...
+
+// HandlerError wraps standard error
+// This error is handled specially by protocol.Run
+// It causes the protocol to return with ErrHandler(err)
+type breakError struct {
+	err error
+}
+
+func BreakError(err error) *breakError {
```
Comment (janos) [inline p2p/protocols/error.go]: Do not return unexported type. Use interface or export breakError.

Constructor should be called `NewBreakError`. Or maybe just `Break`?

File: p2p/protocols/protocol.go [hunk 22]
```diff
@@ -290,9 +226,15 @@ func (p *Peer) run(handler func(ctx context.Context, msg interface{}) error) fun
...
-					case errc <- err:
-					default:
+					var e *breakError
+					if errors.As(err, &e) {
+						select {
+						case errc <- err:
+						default:
+						}
+					} else {
+						// todo: what to do with errors that are not breaking the loop?
```
Comment (janos) [inline p2p/protocols/protocol.go]: I think that logging such errors is enough.

File: p2p/protocols/protocol.go [hunk 23]
```diff
@@ -359,7 +303,7 @@ func (p *Peer) Send(ctx context.Context, msg interface{}) error {
 
 	code, found := p.spec.GetCode(msg)
 	if !found {
-		return errorf(ErrInvalidMsgType, "%v", code)
+		return fmt.Errorf("Invalid message type %v ", code)
```
Comment (janos) [inline p2p/protocols/protocol.go]: Errors should not start with a capital letter. This goes for all error messages below, also.

File: p2p/protocols/protocol.go [hunk 24]
```diff
@@ -494,7 +437,7 @@ func (p *Peer) Handshake(ctx context.Context, hs interface{}, verify func(interf
 			err = ctx.Err()
 		}
 		if err != nil {
-			return nil, errorf(ErrHandshake, err.Error())
+			return nil, fmt.Errorf("%w", err)
```
Comment (janos) [inline p2p/protocols/protocol.go]: No need to wrap error if there is no additional context.

File: p2p/protocols/protocol_test.go [hunk 25]
```diff
@@ -472,19 +472,16 @@ func TestPeer_Run(t *testing.T) {
...
 		rw.eof = true
 		done := make(chan error)
 		go func() {
 			done <- eg.Wait()
 		}()
 		select {
 		case err := <-done:
-			if err != io.EOF {
-				t.Fatal("run returned error")
+			fmt.Println("in done")
```
Comment (janos) [inline p2p/protocols/protocol_test.go]: Remove fmt.Println.

File: pss/pss.go [hunk 26]
```diff
@@ -263,7 +263,14 @@ func (p *Pss) Run(peer *p2p.Peer, rw p2p.MsgReadWriter) error {
 	handle := func(ctx context.Context, msg interface{}) error {
 		return p.handle(ctx, pp, msg)
 	}
-	return pp.Run(handle)
+
+	if err := pp.Run(handle); err != nil {
+		log.Error("Shutting down pss protocol.", "peer", pp, "reason", err)
```
Comment (janos) [inline pss/pss.go]: Error logged and returned.

File: pss/pss.go [hunk 27]
```diff
@@ -423,17 +430,13 @@ func (p *Pss) deregister(topic *message.Topic, hndlr *handler) {
...
-			log.Warn("handler error", "err", err)
-			peer.Drop(fmt.Sprintf("handler error %s", err))
-		}
-	}()
+	pssmsg, ok := msg.(*message.Message)
+	if !ok {
+		return fmt.Errorf("invalid message type: %s", msg)
+	}
+	if err := p.handlePssMsg(ctx, pssmsg); err != nil {
+		return fmt.Errorf("handler error: %w", err)
```
Comment (janos) [inline pss/pss.go]: No need for word `error` in error message.

File: network/stream/stream.go [hunk 28]
```diff
@@ -141,7 +141,13 @@ func (r *Registry) Run(bp *network.BzzPeer) error {
 
 	go sp.InitProviders()
 
-	return sp.Peer.Run(r.HandleMsg(sp))
+	// todo: log peer info appropriately
+	if err := sp.Peer.Run(r.HandleMsg(sp)); err != nil {
+		log.Error("Shutting down stream protocol.", "peer:", sp, "reason", err)
```
Comment (janos) [inline network/stream/stream.go]: If there is an error, both log.Error and log.Info bellow will be printed.

**Review by janos — COMMENTED** [review]
I like the approach with BreakError (even thou I have comments on its implementation).

My comments are mostly on two topics. One is the logging and returning errors at the same time, which should not be done, as the error may be logged twice. The second is related to error messages. They can be simplified or shortened, buy removing words `error`, `erroring`, `err:`, `reason:` as they are errors implicitly. Also errors should not start with a capital letter, or have a form of full sentences. I have left comments on that, but not on all occurrences, as they would be mostly repetitive.

File: bzzeth/bzzeth.go (hunk 11 above)
Comment (pradovic) [reply bzzeth/bzzeth.go]: This (and the next comment) were the things I wanted to discuss.  I see 3 scenarios:
1. Log some context and message here and return nil
2. Log some context here and return a new specific error that will be seen in events in tests
3. Add more context too the error here and return an error. I am not sure if this error will be logged anywhere.

File: p2p/protocols/error.go (hunk 21 above)
Comment (pradovic) [reply p2p/protocols/error.go]: Sounds good. Hm, I was also wondering how to call it. I think `Break` sounds good to me event, I was not sure if something like this has to little info.

File: p2p/protocols/protocol.go (hunk 22 above)
Comment (pradovic) [reply p2p/protocols/protocol.go]: Debug, or maybe even Info? I would probably not log it as error, as it is not breaking the loop? Or maybe log additional context like `non fatal: err`?

File: p2p/protocols/protocol.go (hunk 23 above)
Comment (pradovic) [reply p2p/protocols/protocol.go]: I agree, I did not change this yet, as most of the messages was like this, wanted to brainstorm about the errors first and then change, but I think I have enough info now in the comments about the different errors :)

File: p2p/protocols/protocol_test.go (hunk 25 above)
Comment (pradovic) [reply p2p/protocols/protocol_test.go]: Debug line 🙈

File: bzzeth/bzzeth.go (hunk 11 above)
Comment (janos) [reply bzzeth/bzzeth.go]: Scenario 3. is the one that I think is the correct one. We now have BreakError which is to make the peer drop, and any other error will be at least logged. If the error is logged and returned, it is easy to make the mistake to log the error twice. If the error is just passed, it is up to the caller to decide what to do, and the function that returns the error does not have to worry about it.

File: p2p/protocols/protocol.go (hunk 22 above)
Comment (janos) [reply p2p/protocols/protocol.go]: I think that Warn would be appropriate.

File: bzzeth/bzzeth.go (hunk 11 above)
Comment (pradovic) [reply bzzeth/bzzeth.go]: I agree, from the architecture point of view. I am just worried about the `BreakError`s, which won't be logged in this case, right? This is probably the place that has enough information about what will happen to the specific protocol (ex. bzzeth, stream, etc...).

Comment (janos) [reply bzzeth/bzzeth.go]: If there is a BreakError, then it should be logged as error before the peer is dropped.

Comment (pradovic) [reply bzzeth/bzzeth.go]: The peer is actually dropped when the parent run function returns. This is why I was logging it here, as the drop is actually happening when it's propagated with end of the run function, in this  case  when `BzzEth.Run()` returns. This is why I was logging it here. The error returned by `BzzEth.Run()` is only used for events and testDisconnect. So, I see these options if we don't won't to log and return err message here: 
1. Use `peer.Drop()` on error in `p2p.Protocol` and log BreakErrors there before drop. No need to return specific errors in `p2p.Protocol.Run()` then at all maybe?
2. Log the specific info here, and return different error to be shown in events and drop

I think the main question is, weather we need to return specific errors with `p2p.Protocol.Run()` and weather we want to rely on finishing `Run()` functions to drop a peer or do we want to drop a peer with `peer.Drop()`.

Comment (janos) [reply bzzeth/bzzeth.go]: Cool, as we discussed, option 1. would be a better compromise.

Comment (pradovic) [reply bzzeth/bzzeth.go]: Ok, I have switched back to option 1. I don't like the fact that we lose a bit of control over the async errors. It's hard to check what caused protocol to stop, except that there is a log message. Of course, we could add the functionality in p2p.protocol to remember the drop error, and expose the API to check it.  But maybe we can think about all this more in the future, and leave this as is for now, as we tried both approaches and it should not be to hard to switch back and forth between them.

File: bzzeth/bzzeth.go [hunk 29]
```diff
@@ -418,11 +417,11 @@ DELIVERY:
 	if err == nil && deliveredCnt < total {
 		err := p.Send(ctx, &BlockHeaders{Rid: uint32(msg.Rid)})
 		if err != nil {
-			p.logger.Error("could not send empty BlockHeader", "err", err)
+			return fmt.Errorf("could not send empty BlockHeader: %w", err)
```
Comment (acud) [inline bzzeth/bzzeth.go:420]: i am assuming that here `peer` is the ethereum node that expects the block headers. the send fails here, don't we want to break too?

File: p2p/protocols/protocol.go [hunk 30]
```diff
@@ -326,6 +252,10 @@ func (p *Peer) Drop(reason string) {
 // Returns nil if the active jobs are finished within the timeout duration, or error otherwise.
 func (p *Peer) Stop(timeout time.Duration) error {
 	p.mtx.Lock()
+	if p.running != true {
```
Comment (acud) [inline p2p/protocols/protocol.go]: `if !p.running`

File: p2p/protocols/protocol.go [hunk 31]
```diff
@@ -359,7 +286,7 @@ func (p *Peer) Send(ctx context.Context, msg interface{}) error {
 
 	code, found := p.spec.GetCode(msg)
 	if !found {
-		return errorf(ErrInvalidMsgType, "%v", code)
+		return fmt.Errorf("invalid message type %v ", code)
```
Comment (acud) [inline p2p/protocols/protocol.go:290]: probably better to break here. if the message code is not found - it means the nodes are incompatible and we should disconnect from that node

File: p2p/protocols/protocol.go [hunk 32]
```diff
@@ -459,7 +385,7 @@ func (p *Peer) handleMsg(msg p2p.Msg, handle func(ctx context.Context, msg inter
 // returns the remote handshake and an error
 func (p *Peer) Handshake(ctx context.Context, hs interface{}, verify func(interface{}) error) (interface{}, error) {
 	if _, ok := p.spec.GetCode(hs); !ok {
-		return nil, errorf(ErrHandshake, "unknown handshake message type: %T", hs)
+		return nil, fmt.Errorf("unknown handshake message type: %T", hs)
```
Comment (acud) [inline p2p/protocols/protocol.go:424]: do we break as a result?

File: pss/pss.go [hunk 33]
```diff
@@ -423,18 +424,11 @@ func (p *Pss) deregister(topic *message.Topic, hndlr *handler) {
...
-		}
-		if err := p.handlePssMsg(ctx, pssmsg); err != nil {
-			log.Warn("handler error", "err", err)
-			peer.Drop(fmt.Sprintf("handler error %s", err))
-		}
-	}()
-	return nil
+	pssmsg, ok := msg.(*message.Message)
+	if !ok {
+		return fmt.Errorf("invalid message type %s", msg)
```
Comment (acud) [inline pss/pss.go:429]: not sure if we should break or not. @nolash ?

**Review by acud — APPROVED** [review]
Wonderful work on this PR @pradovic! :clap: :champagne: 
I have just a few minor comments

File: p2p/protocols/protocol.go (hunk 31 above)
Comment (pradovic) [reply p2p/protocols/protocol.go:290]: This is a sync call, as it is not using an event loop, so here we don't need to use a special error. This goes for `Handshake()` calls as well. Only run needs a special `BreakError` in order to break. Does this makes sense?

**Review by janos — APPROVED** [review]
Thanks, LGTM.
